File: Python/mms/MMSRetrieve.py
from response.mms.InboundMMSMessage import InboundMMSMessage
from response.mms.InboundMessageList import InboundMessageList
from response.mms.RetrieveMMSResponse import RetrieveMMSResponse
from response.mms.InboundMessage import InboundMessage
from response.mms.RetrieveMMSMessageResponse import RetrieveMMSMessageResponse
from response.ResourceReference import ResourceReference
from response.mms.MMSMessageReceiptSubscriptionResponse import MMSMessageReceiptSubscriptionResponse
from response.mms.HTTPResponse import HTTPResponse
import json
from foundation.FormParameters import FormParameters
from foundation.JSONRequest import JSONRequest
import email
from response.Attachment import Attachment

class MMSRetrieve:
	"""This class implements the MMS receive group of functions"""
	
	def __init__(self, endpoints, username, password):
		"""default constructor for MMSRetrieve requires definition of OneAPI server endpoints and the username/password for authentication"""
		self.endpoints=endpoints
		self.username=username
		self.password=password
	
	def getEndpoints(self):
		"""getter for endpoints : the network endpoint locations for each of the OneAPI services"""
		return self.endpoints
	
	def setEndpoints(self,endpoints):
		"""setter for endpoints : the network endpoint locations for each of the OneAPI services"""
		self.endpoints=endpoints
	
	def getUsername(self):
		"""getter for username : the username required for authentication with the OneAPI server"""
		return self.username
	
	def setUsername(self,username):
		"""setter for username : the username required for authentication with the OneAPI server"""
		self.username=username
	
	def getPassword(self):
		"""getter for password : the password required for authentication with the OneAPI server"""
		return self.password
	
	def setPassword(self,password):
		"""setter for password : the password required for authentication with the OneAPI server"""
		self.password=password
	
	
	def retrieveMessages(self,registrationId,maxBatchSize):
		"""
		retrieveMessages : retrieve a list of messages sent to your Web application
		Parameters:
		registrationId : the registration ID agreed with the OneAPI operator.
		maxBatchSize : the maximum number of messages to retrieve in this request
		"""
		baseurl=self.endpoints.getRetrieveMMSEndpoint()
		requestProcessor=JSONRequest()
		if '{registrationId}' in baseurl: baseurl=baseurl.replace('{registrationId}',str(registrationId))
		if '{maxBatchSize}' in baseurl: baseurl=baseurl.replace('{maxBatchSize}',str(maxBatchSize))
		rawresponse=requestProcessor.get(baseurl,'application/json', self.username, self.password)
		response=RetrieveMMSResponse()
		if rawresponse is not None and rawresponse.getContent() is not None:
			jsondata=json.loads(rawresponse.getContent())
			if jsondata is not None and jsondata['inboundMessageList'] is not None:
				response.setInboundMessageListJSON(jsondata['inboundMessageList'])
		if rawresponse.getCode() is not None: response.setHTTPResponseCode(rawresponse.getCode())
		if rawresponse.getLocation() is not None: response.setLocation(rawresponse.getLocation())
		if rawresponse.getContentType() is not None: response.setContentType(rawresponse.getContentType())
		return response
	
	def retrieveMessageContent(self,registrationId,messageId,resFormat):
		"""
		retrieveMessageContent : retrieve the full MMS including attachments
		Parameters:
		registrationId : the registration ID agreed with the OneAPI operator
		messageId : is a server-generated message identifier (which can be obtained from the retrieveMessages response)
		resFormat : resFormat=JSON ensures a JSON response content-type
		"""
		baseurl=self.endpoints.getRetrieveMMSMessageEndpoint()
		requestProcessor=JSONRequest()
		if '{registrationId}' in baseurl: baseurl=baseurl.replace('{registrationId}',str(registrationId))
		if '{messageId}' in baseurl: baseurl=baseurl.replace('{messageId}',str(messageId))
		if '{resFormat}' in baseurl: baseurl=baseurl.replace('{resFormat}',str(resFormat))
		rawresponse=requestProcessor.get(baseurl,'application/json', self.username, self.password)
		response=RetrieveMMSMessageResponse()
		if rawresponse is not None and rawresponse.getContent() is not None:
			mimecontent='Content-Type: '+rawresponse.getContentType()+'\r\n'+rawresponse.getContent()
			response=RetrieveMMSMessageResponse()
			response.setAttachment(None)
			mimemsg=email.message_from_string(mimecontent)
			for part in mimemsg.walk():
				contentType=part.get_content_type()
				if 'multipart/mixed' not in contentType:
					filename=part.get_filename()
					payload=part.get_payload()
					if contentType is not None and contentType=='application/json' and filename is not None and filename=='root-fields' and payload is not None:
						jsondata=json.loads(payload)
						if jsondata is not None and jsondata['inboundMessage'] is not None:
							response.setInboundMessage(jsondata['inboundMessage'])
					else:
						attachment=Attachment(None)
						attachment.setContentType(contentType)
						attachment.setName(filename)
						attachment.setData(payload)
						if response.getAttachment()==None: response.setAttachment([])
						response.getAttachment().append(attachment)
		if rawresponse.getCode() is not None: response.setHTTPResponseCode(rawresponse.getCode())
		if rawresponse.getLocation() is not None: response.setLocation(rawresponse.getLocation())
		if rawresponse.getContentType() is not None: response.setContentType(rawresponse.getContentType())
		return response
	
	def subscribeToReceiptNotifications(self,destinationAddress,notifyURL,criteria,notificationFormat,clientCorrelator,callbackData):
		"""
		subscribeToReceiptNotifications : subscribe to notifications of MMS messages sent to your application
		Parameters:
		destinationAddress : the MSISDN, or code agreed with the operator, to which people may send an MMS to your application
		notifyURL : (URL) is your address to which notifications will be sent
		criteria : (string) is case-insensitve text to match against the first word of the message, ignoring any leading whitespace. This allows you to reuse a short code among various applications, each of which can register their own subscription with different criteria.
		notificationFormat : is the content type that notifications will be sent in - for OneAPI v1.0 only JSON is supported.
		clientCorrelator : (string) uniquely identifies this create subscription request. If there is a communication failure during the request, using the same clientCorrelator when retrying the request allows the operator to avoid creating a duplicate subscription.
		callbackData : (string) is a function name or other data that you would like included when the POST is received.
		"""
		baseurl=self.endpoints.getMMSReceiptSubscriptionsEndpoint()
		requestProcessor=JSONRequest()
		formparameters=FormParameters()
		formparameters.put('destinationAddress',destinationAddress)
		formparameters.put('notifyURL',notifyURL)
		formparameters.put('criteria',criteria)
		formparameters.put('notificationFormat',notificationFormat)
		formparameters.put('clientCorrelator',clientCorrelator)
		formparameters.put('callbackData',callbackData)
		postdata=formparameters.encodeParameters()
		rawresponse=requestProcessor.post(baseurl,postdata,'application/json', self.username, self.password)
		response=MMSMessageReceiptSubscriptionResponse()
		if rawresponse is not None and rawresponse.getContent() is not None:
			jsondata=json.loads(rawresponse.getContent())
			if jsondata is not None and jsondata['resourceReference'] is not None:
				response.setResourceReferenceJSON(jsondata['resourceReference'])
		if rawresponse.getCode() is not None: response.setHTTPResponseCode(rawresponse.getCode())
		if rawresponse.getLocation() is not None: response.setLocation(rawresponse.getLocation())
		if rawresponse.getContentType() is not None: response.setContentType(rawresponse.getContentType())
		return response
	
	def cancelReceiptNotifications(self,subscriptionId):
		"""
		cancelReceiptNotifications : stop the subscription to MMS message notifications
		Parameters:
		subscriptionId : subscriptionId of a previously created deliveryNotification subscription
		"""
		baseurl=self.endpoints.getCancelMMSReceiptSubscriptionEndpoint()
		requestProcessor=JSONRequest()
		if '{subscriptionId}' in baseurl: baseurl=baseurl.replace('{subscriptionId}',str(subscriptionId))
		rawresponse=requestProcessor.delete(baseurl,self.username,self.password)
		response=HTTPResponse()
		if rawresponse.getCode() is not None: response.setHTTPResponseCode(rawresponse.getCode())
		if rawresponse.getLocation() is not None: response.setLocation(rawresponse.getLocation())
		if rawresponse.getContentType() is not None: response.setContentType(rawresponse.getContentType())
		return response

File: Ruby/mms/MMSRetrieve.rb
require "response/mms/InboundMMSMessage"
require "response/mms/InboundMessageList"
require "response/mms/RetrieveMMSResponse"
require "response/mms/InboundMessage"
require "response/mms/RetrieveMMSMessageResponse"
require "response/ResourceReference"
require "response/mms/MMSMessageReceiptSubscriptionResponse"
require "response/mms/HTTPResponse"
require 'json'
require 'uri'
require 'net/http'
require 'foundation/FormParameters'
require 'foundation/JSONRequest'
require 'Mail'
require 'response/Attachment'

class MMSRetrieve

	def initialize(endpoints, username, password)
		@endpoints=endpoints
		@username=username
		@password=password
	end
	def getEndpoints
		@endpoints
	end

	def setEndpoints(endpoints)
		@endpoints=endpoints
	end

	def getUsername
		@username
	end

	def setUsername(username)
		@username=username
	end

	def getPassword
		@password
	end

	def setPassword(password)
		@password=password
	end

	
	def retrieveMessages(registrationId,maxBatchSize)
		baseurl=@endpoints.getRetrieveMMSEndpoint()
		requestProcessor=JSONRequest.new()
		if baseurl.index('{registrationId}')!=nil then
			baseurl=baseurl.gsub('{registrationId}',CGI::escape(registrationId.to_s))
		end
		if baseurl.index('{maxBatchSize}')!=nil then
			baseurl=baseurl.gsub('{maxBatchSize}',CGI::escape(maxBatchSize.to_s))
		end
		rawresponse=requestProcessor.get(baseurl,'application/json', @username, @password)
		response=RetrieveMMSResponse.new()
		if (rawresponse!=nil) && (rawresponse.getContent()!=nil)
			jsondata=JSON.parse(rawresponse.getContent())
			if (jsondata!=nil) && (jsondata['inboundMessageList']!=nil) then
				response.setInboundMessageListJSON(jsondata['inboundMessageList'])
			end
		end
		if rawresponse.getCode()!=nil then
			response.setHTTPResponseCode(rawresponse.getCode())
		end
		if rawresponse.getLocation()!=nil then
			response.setLocation(rawresponse.getLocation())
		end
		if rawresponse.getContentType()!=nil then
			response.setContentType(rawresponse.getContentType())
		end
		return response
	end
	
	def retrieveMessageContent(registrationId,messageId,resFormat)
		baseurl=@endpoints.getRetrieveMMSMessageEndpoint()
		requestProcessor=JSONRequest.new()
		if baseurl.index('{registrationId}')!=nil then
			baseurl=baseurl.gsub('{registrationId}',CGI::escape(registrationId.to_s))
		end
		if baseurl.index('{messageId}')!=nil then
			baseurl=baseurl.gsub('{messageId}',CGI::escape(messageId.to_s))
		end
		if baseurl.index('{resFormat}')!=nil then
			baseurl=baseurl.gsub('{resFormat}',CGI::escape(resFormat.to_s))
		end
		rawresponse=requestProcessor.get(baseurl,'application/json', @username, @password)
		response=RetrieveMMSMessageResponse.new()
		if (rawresponse!=nil) && (rawresponse.getContent()!=nil) then
			mimecontent="Content-Type: "+rawresponse.getContentType()+'\r\n\r\n'+rawresponse.getContent()
			mimemsg=Mail.read_from_string(mimecontent.gsub("\\t","\t").gsub("\\n","\n").gsub("\\r","\r").gsub("\\\"","\""))
			response=RetrieveMMSMessageResponse.new()
			attachments=nil
			for part in mimemsg.parts
				contentType=part.content_type.strip
				if (contentType.index(";")!=nil)
					ct=contentType.split(";")
					contentType=ct[0].strip
				end
				filename=nil
				contentDisposition=part.content_disposition
				if (contentDisposition!=nil)
					cdp=contentDisposition.split(";")
					for cvp in cdp
						kv=cvp.split("=")
						if kv[0].strip=="name"
							filename=kv[1].gsub("\"","").strip
						end
					end
				end
				if contentType!='multipart/mixed'
					payload=part.body.to_s
					if (contentType!=nil) && (contentType=='application/json') && (filename!=nil) && (filename=="root-fields") && (payload!=nil)
						jsondata=JSON.parse(payload)
						if (jsondata!=nil) && (jsondata['inboundMessage']!=nil)
							response.setInboundMessage(jsondata['inboundMessage'])
						end
					else
						attachment=Attachment.new
						attachment.setContentType(contentType)
						attachment.setName(filename)
						attachment.setData(payload)
						if attachments==nil
							attachments=Array.new
						end
						attachments[attachments.length]=attachment
					end
				end
			end
			response.setAttachment(attachments)
		end
		if rawresponse.getCode()!=nil then
			response.setHTTPResponseCode(rawresponse.getCode())
		end
		if rawresponse.getLocation()!=nil then
			response.setLocation(rawresponse.getLocation())
		end
		if rawresponse.getContentType()!=nil then
			response.setContentType(rawresponse.getContentType())
		end
		return response
	end
	
	def subscribeToReceiptNotifications(destinationAddress,notifyURL,criteria,notificationFormat,clientCorrelator,callbackData)
		baseurl=@endpoints.getMMSReceiptSubscriptionsEndpoint()
		requestProcessor=JSONRequest.new()
		formparameters=FormParameters.new()
		formparameters.put('destinationAddress',destinationAddress)
		formparameters.put('notifyURL',notifyURL)
		formparameters.put('criteria',criteria)
		formparameters.put('notificationFormat',notificationFormat)
		formparameters.put('clientCorrelator',clientCorrelator)
		formparameters.put('callbackData',callbackData)
		postdata=formparameters.encodeParameters()
		rawresponse=requestProcessor.post(baseurl,postdata,'application/json', @username, @password)
		response=MMSMessageReceiptSubscriptionResponse.new()
		if (rawresponse!=nil) && (rawresponse.getContent()!=nil)
			jsondata=JSON.parse(rawresponse.getContent())
			if (jsondata!=nil) && (jsondata['resourceReference']!=nil) then
				response.setResourceReferenceJSON(jsondata['resourceReference'])
			end
		end
		if rawresponse.getCode()!=nil then
			response.setHTTPResponseCode(rawresponse.getCode())
		end
		if rawresponse.getLocation()!=nil then
			response.setLocation(rawresponse.getLocation())
		end
		if rawresponse.getContentType()!=nil then
			response.setContentType(rawresponse.getContentType())
		end
		return response
	end
	
	def cancelReceiptNotifications(subscriptionId)
		baseurl=@endpoints.getCancelMMSReceiptSubscriptionEndpoint()
		requestProcessor=JSONRequest.new()
		if baseurl.index('{subscriptionId}')!=nil then
			baseurl=baseurl.gsub('{subscriptionId}',CGI::escape(subscriptionId.to_s))
		end
		rawresponse=requestProcessor.delete(baseurl,@username,@password)
		response=HTTPResponse.new()
		if rawresponse.getCode()!=nil then
			response.setHTTPResponseCode(rawresponse.getCode())
		end
		if rawresponse.getLocation()!=nil then
			response.setLocation(rawresponse.getLocation())
		end
		if rawresponse.getContentType()!=nil then
			response.setContentType(rawresponse.getContentType())
		end
		return response
	end

end 
